### PR TITLE
Upgrade to AHC 1.9.11, refactor WSRequestHolder into WSRequest.

### PR DESCRIPTION
**Pull request by wsargent**
Uses AHC signature generator instead of Signpost. There's an edge case where OAuth signing with a POST needs to be handled differently, because the OAuthSignatureGenerator uses getFormParams() and not setBody() for calculating the signature.

The RequestBuilder is also isolated in a single method, so that mutable state is contained.

TODO: NingWSRequest.java needs the same to addQueryParam for OAuth signing.

### CODE REVIEW

**Comment by benmccann** [comment]
@wsargent this failed the sclariform check

**Comment by jroper** [comment]
@wsargent I think a merge and/or rebase failed, I don't know what's going on but there's a few commits in this PR that shouldn't be there.

**Comment by wsargent** [comment]
@jroper will rebase...

**Comment by wsargent** [comment]
Also note this has some crossover with https://github.com/playframework/playframework/issues/3365 -- it may be easier to use OAuthRequestVerifier and have more direct control over signing.

**Comment by wsargent** [comment]
Also see https://github.com/playframework/playframework/pull/2899

**Comment by wsargent** [comment]
and https://github.com/playframework/playframework/pull/3434

File: framework/src/play-java-ws/src/test/scala/play/libs/oauth/OAuthSpec.scala [hunk 1]
```diff
@@ -1,63 +0,0 @@
...
-import play.api.mvc._
-import play.api.test._
-
-import scala.concurrent.Promise
-import play.libs.F
-import play.libs.oauth.OAuth._
-import play.libs.ws.WS
-import play.api.libs.oauth.OAuthRequestVerifier
-
-class OAuthSpec extends PlaySpecification {
```
Comment (jroper) [inline framework/src/play-java-ws/src/test/scala/play/libs/oauth/OAuthSpec.scala]: Should this spec be deleted? Shouldn't we have an end to end check for the Java WS API too?

**Comment by jroper** [comment]
Hey Will,

Thinking this over a little more, we have a request, and generally you would think of a request as an immutable thing.  Then you have the client, which takes a request, and actually makes it.  But, at the same time, we want a fluent API, it's nice to be able to go `ws.url("...").get()`, and this means the request isn't just an immutable object, it's got a client attached to it.

So, not thinking about backwards compatibility or anything like that for a moment, we could do this:

```scala
wsClient.execute(WSRequest(method, url).withHeaders(foo -> bar))
```

But that doesn't look very nice.  One option we could take is this:

```scala
trait WSRequestBuilder[B <: WSRequestBuilder[B]] {
  def withHeaders(headers: (String, String)*): B
  ...
}

case class WSRequest extends WSRequestBuilder[WSRequest](
  headers: Map[String, Seq[String]],
  ...
) {
  def withHeaders(headers: (String, String)*): WSRequest = 
    copy(headers = this.headers ++ headers.toMap)
  ...
}

abstract class WSClientRequest(request: WSRequest) extends WSRequestBuilder[WSClientRequest] {
  protected def withRequest(request: WSRequest): WSClientRequest

  def withHeaders(headers: (String, String)*): WSClientRequest =
    withRequest(request.withHeaders(headers))

  def execute() = Future[WSResponse]
}

class NingClientRequest(request: WSRequest, client: NingClient) extends WSClientRequest {
  protected def withRequest(request: WSRequest) = new NingClientRequest(request, client)
  def execute() = { ... }
}
```

This means the signature calculator can just be `WSRequest -> WSRequest`, we keep our request object as a case class (which would be great for implementing a cache for example), plus we also get our fluent API, and the API to modifying `WSRequest` is the same to modifying the client request fluent API.  What do you think?

**Comment by wsargent** [comment]
So you'd have a WSRequest, containing all the user data in an implementation independent format, and you'd have a NingClientRequest that you can call execute on.

I like the idea, but this does mean that you would have a harder time setting client specific per-request properties -- although, the way that we do this now is by setting ws.ning properties on the client, so if they really want different settings they can set up multiple clients with a smaller thread pools.

**Comment by jroper** [comment]
@wsargent if we wanted to allow setting client specific properties, we could do this:

```scala
class NingClientRequest(request: WSRequest, ningRequest: NingRequest, client: NingClient) extends WSClientRequest {
  protected def withRequest(request: WSRequest) = new NingClientRequest(request, client)

  def withNingSpecificProperty(p: String): NingClientRequest = new NingClientRequest(request, ningRequest.withNingSpecificProperty(p), client)

  def execute() = { ... }
}
```

Or something like that.

**Comment by wsargent** [comment]
Also relates to https://github.com/playframework/playframework/pull/3620

**Comment by wsargent** [comment]
routes compiler is failing:

```
java.lang.RuntimeException: Failed tests:
	routes-compiler-plugin / injected-routes-compilation
	routes-compiler-plugin / routes-compilation

	at scala.sys.package$.error(package.scala:27)
	at sbt.test.ScriptedTests$.runAll(ScriptedTests.scala:128)
	at sbt.test.ScriptedTests$.run(ScriptedTests.scala:122)
	at sbt.test.ScriptedTests$.run(ScriptedTests.scala:114)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at sbt.ScriptedPlugin$$anonfun$scriptedTask$1$$anonfun$apply$1$$anonfun$apply$2.apply(ScriptedPlugin.scala:44)
	at sbt.ScriptedPlugin$$anonfun$scriptedTask$1$$anonfun$apply$1$$anonfun$apply$2.apply(ScriptedPlugin.scala:40)
	at scala.Function1$$anonfun$compose$1.apply(Function1.scala:47)
	at sbt.$tilde$greater$$anonfun$$u2219$1.apply(TypeFunctions.scala:42)
	at sbt.std.Transform$$anon$4.work(System.scala:64)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1$$anonfun$apply$1.apply(Execute.scala:237)
	at sbt.ErrorHandling$.wideConvert(ErrorHandling.scala:18)
	at sbt.Execute.work(Execute.scala:244)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:237)
	at sbt.Execute$$anonfun$submit$1.apply(Execute.scala:237)
	at sbt.ConcurrentRestrictions$$anon$4$$anonfun$1.apply(ConcurrentRestrictions.scala:160)
	at sbt.CompletionService$$anon$2.call(CompletionService.scala:30)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
[error] (SBT-Plugin/*:scripted) Failed tests:
[error] 	routes-compiler-plugin / injected-routes-compilation
[error] 	routes-compiler-plugin / routes-compilation
```

File: framework/src/play-ws/src/main/scala/play/api/libs/ws/WS.scala [hunk 2]
```diff
@@ -96,8 +96,8 @@ object WS {
   @deprecated("Please use play.api.libs.ws.WSResponse", "2.3.0")
   type Response = play.api.libs.ws.WSResponse
 
-  @deprecated("Please use play.api.libs.ws.WSRequestHolder", "2.3.0")
-  type WSRequestHolder = play.api.libs.ws.WSRequestHolder
+  @deprecated("Please use play.api.libs.ws.WSRequest", "2.3.0")
+  type WSRequestHolder = play.api.libs.ws.WSRequest
```
Comment (jroper) [inline framework/src/play-ws/src/main/scala/play/api/libs/ws/WS.scala]: Anything that was deprecated in 2.3 we can delete altogether.

File: framework/src/play-java-ws/src/main/java/play/libs/ws/WSRequestHolder.java [hunk 3]
```diff
@@ -1,124 +0,0 @@
...
-
-import com.fasterxml.jackson.databind.JsonNode;
-import play.libs.F;
-
-import java.io.File;
-import java.io.InputStream;
-import java.util.Collection;
-import java.util.Map;
-
-public interface WSRequestHolder {
```
Comment (jroper) [inline framework/src/play-java-ws/src/main/java/play/libs/ws/WSRequestHolder.java]: Let's keep this as an empty interface and put a deprecation notice on it, for documentation purposes.

**Comment by wsargent** [comment]
Also fixes https://github.com/AsyncHttpClient/async-http-client/issues/668

File: framework/src/play-ws/src/main/scala/play/api/libs/ws/WS.scala (hunk 2 above)
Comment (wsargent) [inline framework/src/play-ws/src/main/scala/play/api/libs/ws/WS.scala]: Will do.

File: framework/src/play-java-ws/src/main/java/play/libs/ws/WSRequestHolder.java (hunk 3 above)
Comment (wsargent) [inline framework/src/play-java-ws/src/main/java/play/libs/ws/WSRequestHolder.java]: Okay.

File: framework/project/Dependencies.scala [hunk 4]
```diff
@@ -277,7 +277,7 @@ object Dependencies {
 
   val playWsDeps = Seq(
     guava,
-    "com.ning" % "async-http-client" % "1.8.15"
+    "com.ning" % "async-http-client" % "1.9.11"
   ) ++ Seq("signpost-core", "signpost-commonshttp4").map("oauth.signpost" % _  % "1.2.1.2") ++
```
Comment (jroper) [inline framework/project/Dependencies.scala]: If we're using ning to do OAuth signing, does that mean we can remove this dependency yet?

Comment (wsargent) [inline framework/project/Dependencies.scala]: ALMOST.  There's one tiny place left: the OAuth case class that gets request tokens.  I wasn't sure what to do in this case at the time, but I can go back and clean it out.

https://github.com/wsargent/playframework/blob/upgrade-ahc-1.9.x/framework/src/play-ws/src/main/scala/play/api/libs/oauth/OAuth.scala#L22

Comment (jroper) [inline framework/project/Dependencies.scala]: Right, so we just need to implement the token retrieval.  That shouldn't be part of this change.

**Comment by yardenr** [comment]
What version can i use with the ning upgrade?

**Comment by wsargent** [comment]
@yardenr Sorry, I'm not clear.  What version of what?

**Comment by megazord** [comment]
@yardenr 

This PR was merged after play 2.4.0-M3. It will included in the next release.

**Comment by yardenr** [comment]
@wsargent @megazord 
Thanks a lot.
